typing/printtyp.ml

```diff
@@ -413,22 +350,31 @@ let rewrite_double_underscore_paths env p =
   else
     rewrite_double_underscore_paths env p
 
-let rec tree_of_path namespace = function
+let rec tree_of_path ?(disambiguation=true) namespace p =
+  let namespace = if disambiguation then namespace else Other in
+  match p with
   | Pident id ->
       Oide_ident (ident_name namespace id)
-  | Pdot(_, s) as path when non_shadowed_pervasive path ->
-      Oide_ident (Naming_context.pervasives_name namespace s)
+  | Pdot(_, s) as path when non_shadowed_stdlib namespace path ->
+      Oide_ident (Out_name.create s)
   | Pdot(Pident t, s)
     when namespace=Type && not (Path.is_uident (Ident.name t)) ->
       (* [t.A]: inline record of the constructor [A] from type [t] *)
       Oide_dot (Oide_ident (ident_name Type t), s)
   | Pdot(p, s) ->
-      Oide_dot (tree_of_path Module p, s)
+      Oide_dot (tree_of_path ~disambiguation Module p, s)
   | Papply(p1, p2) ->
-      Oide_apply (tree_of_path Module p1, tree_of_path Module p2)
+      let t1 = tree_of_path ~disambiguation Module p1 in
+      let t2 = tree_of_path ~disambiguation Module p2 in
+      Oide_apply (t1, t2)
+
+let tree_of_path ?disambiguation namespace p =
+  tree_of_path ?disambiguation namespace
+    (rewrite_double_underscore_paths !printing_env p)
 
-let tree_of_path namespace p =
-  tree_of_path namespace (rewrite_double_underscore_paths !printing_env p)
+let best_tree_of_type_path p p' =
+  if p == p' then tree_of_path Type p'
```

Comment:
I'm nervous again here. For me physical equality should be used as an optimization to shortcut a more expansive (but more meaningful) equality check. When you do this, you know the physical check is always an optimization that can b removed without changing the behavior -- if we remove the short path, we end up in the equality check that gives the same result. It is not clear that it is the case here.

Reply:
The simplest fix here is probably to ensure that the `short-path` function returns directly the information that it has returned an alternative path or not.